src/happy/oauth2_capture_redirect.clj

```diff
@@ -14,19 +14,67 @@
     (catch Exception _
       nil)))
 
-(defn wait-for-redirect [{:as config :keys [redirect_uri redirect_uris]}]
+(defn redirect-uri->port
+  "Opening a browser and redirecting to a local server
+  only makes sense for localhost redirect-uris.
+
+  Known limitation: Google does not support no ipv6 [::1] with random ports"
+  [uri]
+  (let [[match protocol host _ port path] (re-find #"^(http://)(localhost|127.0.0.1)(:(\d+))?(/.*)?$" uri)
+        port (maybe-parse-int port)]
+    (if (not match)
+      (throw (RuntimeException. "URI doesn't match pattern"))
+      (or port
+          0)))) ;; Jetty's magic value 0 = random port
+
+(defn add-port
+  [uri random-port]
+  (let [[match protocol host _ port path] (re-find #"^(http://)(localhost|127.0.0.1)(:(\d+))?(/.*)?$" uri)]
+    (if (not match)
+      (throw (RuntimeException. "URI doesn't match pattern"))
+      (str protocol host ":" random-port path))))
+
+(defn query-string->params
+  "Converts an URI query-string a=b&c=d&e=f&...
+  into a param map {:a \"b\" :c \"d\" :e \"f\"}."
+  [q]
+  (let [split-by-& #(str/split % #"&")
+        split-by-= #(str/split % #"=")
+        pairs (map split-by-= (split-by-& q))
+        parameters (into {} (for [[k v] pairs]
+                              [(keyword k) v]))]
+    parameters))
+
+(defn wait-for-redirect
+  "wait-for-redirect takes a callback function and a Google client
+  config. The callback function will be called, when the jetty server
+  has been started. It receives a single argument with the effective
+  URI, i.e. the URI with the correct (maybe random) port."
+  [on-server-started-callback
```

Comment:
Of all possible designs, I decided for a callback function design. Once understood, it's easy to use. Also it allows for use without blocking, e.g. in tests.

Alternatives would have been:
* composite return value: return effective URI together with the promise and let the user deal with everything
* passing some mutable ref as parameter to capture the return values

Reply:
This is a breaking change to `wait-for-redirect`. I considered this function as quasi-internal. That's something we may want to reconsider (and e.g. provide 2 different signatures for that function: 1 legacy compat and 1 with callback).

Reply:
Hi @JohannesFKnauf,
In my opinion it's more straightforward to interleave them
( see proposal https://github.com/timothypratley/happygapi/pull/8 )